PlasmaControl/DESC
# Testing

Upgrades to testing:

- Add full testing of "precise_QH" optimization (broken into stages for parallelization)
- Add tests to compare different methods of computing vacuum heliotron (fixed current vs minimizing J)
- Add testing for python 3.10
- Remove testing for python 3.7
- Mark tests as unit or regression 
  - **NOTE**: all tests must be marked with either `@pytest.mark.unit` or `@pytest.mark.regression` - unmarked tests will not be run as part of the CI testing
  - `regression` tests are ones that test large portions of the code against VMEC/STELLOPT/SIMSOPT base cases - these may only be run occasionally.
  - `unit` tests are for smaller tests that only test certain functionality - these will be run every time.
  - `unit` tests should make limited use of fixtures, currently only `DSHAPE_current`, `SOLOVEV`, and `HELIOTRON_vac` are used in the unit tests.
- `regression` tests are only run on Python 3.9
- `unit` tests are run on Python 3.8 and 3.10
- Parallelize tests on CI using [pytest-split](https://github.com/jerry-git/pytest-split) - currently split into 4 groups for regression tests and 3 groups for unit tests. This can be tweaked in the future to more evenly distribute the load
- Get rid of use of ``unittest`` in favor of always using ``pytest``

Resolves #265
